### PR TITLE
Revolution

### PR DESCRIPTION
**Pull request by asartori86**

### CODE REVIEW

**Comment by asartori86** [comment]
```
template <int dim, int spacedim>
template <typename EnergyType, typename ResidualType>
void
ProblemTemplate<dim,spacedim>::
set_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
                           FEValuesCache<dim,spacedim> &fe_cache,
                           std::vector<EnergyType> &energies,
                           std::vector<std::vector<ResidualType> > &local_residuals,
                           bool compute_only_system_matrix) const
{

  const FEValuesExtractors::Vector displacement(0);

  ////////// conservative section
  //
  EnergyType et = 0; // dummy number to define the type of variables
  this->reinit (et, cell, fe_cache);
  auto &Fs = fe_cache.get_deformation_gradients("solution", "Fu", displacement, et);

  ////////// non-conservative section
  //
  ResidualType rt = 0;
  this->reinit (rt, cell, fe_cache);
  auto &us = fe_cache.get_values("solution", "u", displacement, rt);


  ////////// common variables
  //
  auto &fev = fe_cache.get_current_fe_values();
  const unsigned int n_q_points = us.size();
  auto &JxW = fe_cache.get_JxW_values();


  for (unsigned int q=0; q<n_q_points; ++q)
    {
      ///////////////////////// energetic contribution
      auto &F = Fs[q];
      auto C = transpose(F)*F;

      auto Ic = trace(C);
      auto J = determinant(F);
      auto lnJ = std::log (J);

      auto psi = (mu/2.)*(Ic-dim) - mu*lnJ + (lambda/2.)*(lnJ)*(lnJ);

      energies[0] += (psi)*JxW[q];

      ////////////////////////// residual formulation
      auto &u = us[q];
      for (unsigned int i=0; i<fev.dofs_per_cell(); ++i)
        {
          auto v = fev[displacement] .value(i,q); // test function

          local_residuals[0] -= 0.1*v*u; // non-conservative load

          // matrix[0] is assumed to be the system matrix
          // other matrices are either preconditioner or
          // auxiliary matrices needed to build it.
          //
          // if this function is called to evalute the
          // residual we do not need to assemble them
          // so we guard them
          if (!compute_only_system_matrix)
            local_residuals[1] += v*u;
        }
    }

}
```

**Comment by asartori86** [comment]
che dite vi piacerebbe così? contributo energetico e formulazione residuo insieme nella stessa funzione? @luca-heltai @ESeNonFossiIo @alucantonio ?

**Comment by alucantonio** [comment]
secondo me può essere utile. Mi viene in mente almeno un caso: un corpo elastico (contributo energetico) soggetto a carichi non conservativi (residuo).

**Comment by ESeNonFossiIo** [comment]
A me piace. Mi sembra più snello come codice.

**Comment by asartori86** [comment]
This is ready for review. I left out for different PRs 
- the definition of the old solution, which, in my opinion, should be done inside pidomus (I left it inside interface)
- solver shold restart 
- definition of a steady_state_solver and a time_dependent solver

**Comment by asartori86** [comment]
Oh yes, I forgot to mention that for sure I have put not enough documentation :)

File: include/base_interface.h [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+ *
+ * Goal: provide a derivable interface to solve a particular
+ *       PDEs problem (time depending, first-order, non linear).
+ *
+ * Usage: This class requires some arguments related to the setting
+ *        of the problem: finite elements, boundary conditions,
+ *        and initial conditions.
+ *        Moreover, it helps to write the system matrix and
+ *        the preconditioner matrix.
+ * Varibles:
```
Comment (luca-heltai) [inline include/base_interface.h]: Variables

File: include/base_interface.h [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+  /**
+   * Initialize all data required for the system
+   *
+   * This function is used to initialize the varibale AnyData @p d
+   * that contains all data of the problem (solutions, DoF, quadrature
+   * points, solutions vector, etc ).
+   * It takes as argument the number of DoF per cell @p dofs_per_cell,
+   * the number of quadrature points @p n_q_points, the number of
+   * quadrature points per face @p n_face_q_points, the reference to
+   * solutions vectors @p sol and the reference to the AnyData @p d.
```
Comment (luca-heltai) [inline include/base_interface.h]: This comment should be updated. n_q_points, n_face_q_points, etc. should all disappear from the comment

File: include/base_interface.h [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                                        const std::vector<shared_ptr<typename LAC::BlockMatrix> >,
+                                        LinearOperator<typename LAC::VectorType> &,
+                                        LinearOperator<typename LAC::VectorType> &) const;
+
+
+  shared_ptr<Mapping<dim,spacedim> > get_mapping() const;
+
+  virtual unsigned int set_mapping_degree() const;
+
+  virtual UpdateFlags get_face_update_flags() const;
```
Comment (luca-heltai) [inline include/base_interface.h:149]: Shouldn't we have also in this guy a number indicating the index of the matrix we are assembling for?

File: include/base_interface.h [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+  virtual UpdateFlags get_matrices_update_flags() const;
```
Comment (luca-heltai) [inline include/base_interface.h]: same here

Comment (luca-heltai) [inline include/base_interface.h]: This is now called `get_matrices_update_flags`, while the above is called `get_face_update_flags`. I think the correct name here should be `get_cell_update_flags(unsigned int index=0)` and `get_face_update_flags(unsigned int index=0)`

File: include/base_interface.h [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+  template<typename Number>
+  void reinit(const Number &alpha,
+              const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+              const unsigned int face_no,
+              FEValuesCache<dim,spacedim> &fe_cache) const;
+
+
+
+  const Table<2,DoFTools::Coupling> &get_matrix_coupling(const unsigned int &i) const;
+  virtual unsigned int get_number_of_matrices() const;
```
Comment (luca-heltai) [inline include/base_interface.h]: n_matrices?

File: include/base_interface.h [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+   * @code
+   * void set_matrices_coupling(std::vector<std::string &couplings) const
+   * {
+   *   // suppose we have 2 matrices (system and preconditioner)
+   *   // we are solving incompressible Stokes equations
+   *   couplings[0] = "1,1;1,0"; // first block is the velocity, the second is the pressure
+   *   couplings[1] = "1,0;0,1";
+   * }
+   */
+  virtual void set_matrices_coupling(std::vector<std::string> &couplings) const;
```
Comment (luca-heltai) [inline include/base_interface.h]: matrices here is used as an adjective: it should be singular. `set_matrix_couplings`

File: include/base_interface.h [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+  const Table<2,DoFTools::Coupling> &get_matrix_coupling(const unsigned int &i) const;
+  virtual unsigned int get_number_of_matrices() const;
+
+protected:
+
+  void build_couplings();
+
+  /**
+   * Define the coupling of each matrix among its blocks.  By default it
+   * sets a fully coupling among each block.  If you want to specify the
```
Comment (luca-heltai) [inline include/base_interface.h:193]: a full coupling

File: include/interfaces/compressible_neo_hookean_iterative.h [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      auto lnJ = std::log (J);
+
+      EnergyType psi = (mu/2.)*(Ic-dim) - mu*lnJ + (lambda/2.)*(lnJ)*(lnJ);
+
+      energies[0] += (psi)*JxW[q];
+
+      if (!compute_only_system_matrix)
+        energies[1] += 0.5*u*u*JxW[q];
+
+      (void)compute_only_system_matrix;
```
Comment (luca-heltai) [inline include/interfaces/compressible_neo_hookean_iterative.h]: This is not required here. You use the variable above.

File: include/interfaces/template.h [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+  // this function defines the order of the mapping used when
+  // Dirichlet boundary conditions are applied, when the Initial
+  // solution is interpolated, when the solution vector is stored in
+  // vtu format and when the the error_from_exact is performed.
+  // By default it returns 1;
+  //
+  // unsigned int set_order_of_mapping () const;
+
+  // set the number of matrices to be assembled
```
Comment (luca-heltai) [inline include/interfaces/template.h:43]: set or get? This only returns a number. It should be `n_matrices`...

File: include/interfaces/template.h [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  //
+  // void set_matrices_coupling (std::vector<std::string> > &couplings) const
+  // {
+  //   couplings[0]="1";
+  //   couplings[1]="1";
+  // }
+
+
+  template <typename EnergyType, typename ResidualType>
+  void set_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
```
Comment (luca-heltai) [inline include/interfaces/template.h:64]: again... set or get? :) 

`set_xxx`: call this function to set an internal variable to the given argument
`get_xxx`: call this function to get the value of an internal variable
`compute_xxx`/`assemble_xxx`: call this function to perform computations

File: include/pidomus.h [hunk 11]
```diff
@@ -185,12 +173,7 @@ class piDoMUS : public ParameterAcceptor, public SundialsInterface<typename LAC:
   shared_ptr<DoFHandler<dim, spacedim> >          dof_handler;
 
   ConstraintMatrix                          constraints;
```
Comment (luca-heltai) [inline include/pidomus.h:215]: We should have a vector of constraints...

File: include/pidomus.h [hunk 12]
```diff
@@ -203,11 +186,9 @@ class piDoMUS : public ParameterAcceptor, public SundialsInterface<typename LAC:
...
   mutable TimerOutput                               computing_timer;
 
-  const unsigned int n_aux_matrices;
-  std::vector<shared_ptr<typename LAC::BlockMatrix> >  aux_matrix;
-  std::vector<shared_ptr<typename LAC::BlockSparsityPattern> > aux_matrix_sp;
-
-
+  const unsigned int n_matrices;
+  std::vector<shared_ptr<typename LAC::BlockMatrix> >  matrices;
+  std::vector<shared_ptr<typename LAC::BlockSparsityPattern> > matrices_sp;
```
Comment (luca-heltai) [inline include/pidomus.h]: matrix_sparsities

File: include/interfaces/compressible_neo_hookean_iterative.h [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+#include <deal.II/lac/packaged_operation.h>
+#include <deal.II/lac/solver_cg.h>
+
+
+template <int dim, int spacedim>
+class CompressibleNeoHookeanInterface : public Interface<dim,spacedim,dim, CompressibleNeoHookeanInterface<dim,spacedim> >
+{
+  typedef TrilinosWrappers::MPI::BlockVector BVEC;
+  typedef TrilinosWrappers::MPI::Vector VEC;
+  typedef TrilinosWrappers::BlockSparseMatrix MAT;
```
Comment (luca-heltai) [inline include/interfaces/compressible_neo_hookean_iterative.h]: use `LAC::MatrixType` and `LAC::VectorType`. To access subblock vector types: `LAC::VectorType::BlockType`

File: include/base_interface.h [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+   */
+  virtual void initialize_data(const typename LAC::VectorType &solution,
+                               const typename LAC::VectorType &solution_dot,
+                               const double t,
+                               const double alpha) const;
+
+
+  /**
+   * Definition of energies and residuals
+   */
```
Comment (luca-heltai) [inline include/base_interface.h:101]: `assemble_energies_and_residuals`

File: include/base_interface.h [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+  virtual void get_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                          FEValuesCache<dim,spacedim> &,
+                                          std::vector<Sdouble> &energies,
+                                          std::vector<std::vector<double> > &local_residuals,
+                                          bool compute_only_system_matrix) const;
+
+  /**
+   * Definition of energies and residuals
+   */
+  virtual void get_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
```
Comment (luca-heltai) [inline include/base_interface.h]: same

File: include/base_interface.h [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+
+  /**
+   * This function can be overloaded to directly implement the local
+   * matrices (i.e. as it is usally done in standard FE codes)
+   */
+  virtual void assemble_local_matrices (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+                                        FEValuesCache<dim,spacedim> &scratch,
+                                        CopyData &data) const;
+
+  virtual void get_local_system_residual (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
```
Comment (luca-heltai) [inline include/base_interface.h]: assemble_local_system_residual

File: include/base_interface.h [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+                                          bool compute_only_system_matrix) const;
+
+  /**
+   * Definition of energies and residuals
+   */
+  virtual void get_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                          FEValuesCache<dim,spacedim> &,
+                                          std::vector<SSdouble> &energies,
+                                          std::vector<std::vector<Sdouble> > &local_residuals,
+                                          bool compute_only_system_matrix) const;
```
Comment (luca-heltai) [inline include/base_interface.h]: compute_only_system_matrix => compute_only_system_terms

File: include/base_interface.h [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+   * This function is used to assemble linear operators related
+   * to the problem.
+   */
+  virtual void compute_system_operators(const DoFHandler<dim,spacedim> &,
+                                        const std::vector<shared_ptr<typename LAC::BlockMatrix> >,
+                                        LinearOperator<typename LAC::VectorType> &,
+                                        LinearOperator<typename LAC::VectorType> &) const;
+
+
+  shared_ptr<Mapping<dim,spacedim> > get_mapping() const;
```
Comment (luca-heltai) [inline include/base_interface.h]: virtual const Mapping<dim,spacedim> & get_mapping() const;

File: include/base_interface.h [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+
+  virtual unsigned int set_mapping_degree() const;
```
Comment (luca-heltai) [inline include/base_interface.h]: remove

File: include/base_interface.h [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+  mutable double old_t;
+
+  mutable double alpha;
+  unsigned int dofs_per_cell;
+  unsigned int n_q_points;
+  unsigned int n_face_q_points;
+
+  mutable unsigned int n_matrices;
+
+  std::vector<Table<2,DoFTools::Coupling> > matrices_coupling;
```
Comment (luca-heltai) [inline include/base_interface.h]: matrix_couplings

**Comment by luca-heltai** [comment]
One of the issues could be addressed by mathLab/deal2lkit#244

File: include/base_interface.h [hunk 21]
```diff
@@ -0,0 +1,280 @@
+/**
+ * Base Interface
+ *
+ * Goal: provide a derivable interface to solve a particular
+ *       PDEs problem (time depending, first-order, non linear).
```
Comment (luca-heltai) [inline include/base_interface.h]: PDE problem (PDE è un aggettivo di problem => singular)

File: include/base_interface.h [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+
+  /**
+   * Constructor. It takes the name of the subsection within the
+   * parameter file, the finite element used to discretize the system,
+   * the name of the components and a string were the block of
+   * differential and algebraic components are specified.
+   */
+  BaseInterface(const unsigned int &n_comp,
+                const unsigned int &n_matrices,
+                const std::string &name="",
```
Comment (luca-heltai) [inline include/base_interface.h]: Nell'articolo di deal2lkit abbiamo dichiarato che il primo argomento delle classi derivate da ParameterAcceptor è sempre il nome. Manteniamo questa "tradizione". Metti prima il nome.

File: include/base_interface.h [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+
+  void init ();
+
+
+  virtual void declare_parameters (ParameterHandler &prm);
+
+  virtual void parse_parameters_call_back() {};
+
+  /**
+   * returns the vector @p of differential blocks
```
Comment (luca-heltai) [inline include/base_interface.h]: Return

Comment (luca-heltai) [inline include/base_interface.h]: All functions documentation should be with an imperative form, following deal.II style: Return this. Make that. Do this.

File: include/base_interface.h [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+   */
+  const std::vector<unsigned int> get_differential_blocks() const;
+
+  /**
+   * This function is used to modify triangulation using boundary_id or manifold_id.
+   * In the case a signal is required, this is the function to modify.
```
Comment (luca-heltai) [inline include/base_interface.h]: Postprocess the newly generated triangulation.

File: include/base_interface.h [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+   * Definition of energies and residuals
+   */
+  virtual void assemble_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                               FEValuesCache<dim,spacedim> &,
+                                               std::vector<Sdouble> &energies,
+                                               std::vector<std::vector<double> > &local_residuals,
+                                               bool compute_only_system_terms) const;
+
+  /**
+   * Definition of energies and residuals
```
Comment (luca-heltai) [inline include/base_interface.h]: Assemble energies and residuals.

File: include/base_interface.h [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+   */
+  virtual void assemble_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                               FEValuesCache<dim,spacedim> &,
+                                               std::vector<SSdouble> &energies,
+                                               std::vector<std::vector<Sdouble> > &local_residuals,
+                                               bool compute_only_system_terms) const;
+
+  /**
+   * This function can be overloaded to directly implement the local
+   * matrices (i.e. as it is usally done in standard FE codes)
```
Comment (luca-heltai) [inline include/base_interface.h]: Assemble local matrices. The default implementation calls BaseInterface::assemble_energies_and_residuals and creates the local matrices by performing automatic differentiation on the results.

File: include/base_interface.h [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+   */
+  virtual void assemble_local_matrices (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+                                        FEValuesCache<dim,spacedim> &scratch,
+                                        CopyData &data) const;
+
+  virtual void assemble_local_system_residual (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
```
Comment (luca-heltai) [inline include/base_interface.h:131]: Assemble the local system residual associated to the given cell. 

This function is called to evaluate the local system residual at each Newton iteration.

**Pull request merged**